prereise/gather/griddata/hifld/data_process/topology.py

```diff
@@ -40,6 +41,40 @@ def min_dist_of_2_conn_comp(nodes1, nodes2, dist_metric, memory_efficient=False)
     return min_dist, res_v1, res_v2
 
 
+def min_dist_of_2_conn_comp_kdtree(nodes1, nodes2, **kwargs):
+    """Calculate the minimum distance of two connected components using KDTree.
+
+    :param pandas.DataFrame nodes1: a data frame contains all the nodes of the first
+        connected component with index being node ID and two columns being latitude
+        and longitude in order.
+    :param pandas.DataFrame nodes2: a data frame contains all the nodes of the second
+        connected component with index being node ID and two columns being latitude
+        and longitude in order.
+    :param \\*\\*kwargs: optional arguments for KDTree query.
+    :return: (*tuple*) -- a tuple contains three elements, the minimum distance
+        follows by a pair of nodes from the two connected components respectively
+        in order
+
+    .. note:: The minimum distance returned by this function depends on the
+        configuration of KDTree by the user specified keyword arguments, Euclidean
+        distance is used by default. See
+        https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.query.html#scipy.spatial.KDTree.query
+        for more information.
+    """
+    if len(nodes1) < len(nodes2):
+        nodes1, nodes2 = nodes2, nodes1
+    tree = KDTree([ll2uv(p[1], p[0]) for p in nodes1.values])
+    shortest_link = nodes2.apply(
+        lambda x: tree.query(ll2uv(*x[::-1]), **kwargs), axis=1
+    )
+    min_dist, sub1_ind = shortest_link.min()
+    sub1 = nodes1.index[sub1_ind]
+    sub2 = shortest_link.loc[shortest_link == (min_dist, sub1_ind)].index[0]
+    return (
+        (min_dist, sub2, sub1) if len(nodes1) < len(nodes2) else (min_dist, sub1, sub2)
+    )
```

Comment:
After this block...
```python
if len(nodes1) < len(nodes2):
     nodes1, nodes2 = nodes2, nodes1
```
...wont `len(nodes1)` always be >= `len(nodes2)`? In the current test, the length of the test connected components are equal, so that first block doesn't kick in either way and the output indices are properly reversed if we reverse the inputs, but if we add another entry to one of the connected components, then the function returns identical outputs even if the ordering of the inputs is switched.

Reply:
Yeah good catch, this is a bug. The intention here is creating the tree using the big island and looping through the nodes in the small island to query to tree. So we would like `nodes1` always represents the big island. Hence, in the situation we swap, we will need to swap the return value to keep output consistent.

Reply:
Fixed in [d03295b](https://github.com/Breakthrough-Energy/PreREISE/pull/228/commits/d03294b40c6c1aee36b1e259296dbaec0364e8c9) and one more test to cover this track is added [65a2557](https://github.com/Breakthrough-Energy/PreREISE/pull/228/commits/65a2557ae356d27522379c83309c75f5f4102192).